File: plugins/examples/buffer_query_demo.lua
-- Buffer Query API Demo Plugin
-- Demonstrates the new buffer query functions in Phase 2

editor.register_command({
    name = "Query Demo: Show Buffer Info",
    description = "Display information about the current buffer",
    action = "show_buffer_info",
    contexts = {"normal"},
    callback = function()
        local buffer_id = editor.get_active_buffer_id()
        local info = editor.get_buffer_info(buffer_id)

        if info then
            local msg = string.format("Buffer %d: %s (%s, %d bytes)",
                info.id,
                info.path ~= "" and info.path or "[No Name]",
                info.modified and "modified" or "saved",
                info.length
            )
            editor.set_status(msg)
        else
            editor.set_status("No buffer info available")
        end
    end
})

editor.register_command({
    name = "Query Demo: Show Cursor Position",
    description = "Display cursor position and selection info",
    action = "show_cursor_info",
    contexts = {"normal"},
    callback = function()
        local cursor = editor.get_primary_cursor()

        if cursor then
            local msg
            if cursor.selection then
                msg = string.format("Cursor at %d, selection: %d-%d (%d chars)",
                    cursor.position,
                    cursor.selection.start,
                    cursor.selection["end"],
                    cursor.selection["end"] - cursor.selection.start
                )
            else
                msg = string.format("Cursor at byte position %d (no selection)", cursor.position)
            end
            editor.set_status(msg)
        else
            editor.set_status("No cursor info available")
        end
    end
})

editor.register_command({
    name = "Query Demo: Count All Cursors",
    description = "Display the number of active cursors",
    action = "count_cursors",
    contexts = {"normal"},
    callback = function()
        local cursors = editor.get_all_cursors()
        editor.set_status(string.format("Active cursors: %d", #cursors))
    end
})

-- NOTE: get_buffer_content() and get_line() are intentionally NOT provided
-- because they would materialize large portions of the buffer into memory,
-- defeating the editor's streaming architecture for huge files (GB+).
--
-- Instead, use the render-line hook to access line content efficiently:
--
-- editor.on("render-line", function(args)
--     local line_content = args.content
--     local line_number = args.line_number
--     -- Process visible lines here
-- end)
--
-- For file-level operations, use external tools via editor.spawn():
--
-- editor.spawn("wc", {"-l", filepath}, function(stdout, stderr, exit_code)
--     local line_count = tonumber(stdout:match("%d+"))
--     editor.set_status(string.format("File has %d lines", line_count))
-- end)

editor.register_command({
    name = "Query Demo: List All Buffers",
    description = "Show count of open buffers",
    action = "list_all_buffers",
    contexts = {"normal"},
    callback = function()
        local buffers = editor.list_buffers()
        local modified_count = 0

        for _, buf in ipairs(buffers) do
            if buf.modified then
                modified_count = modified_count + 1
            end
        end

        editor.set_status(string.format("Open buffers: %d (%d modified)", #buffers, modified_count))
    end
})

editor.register_command({
    name = "Query Demo: Show Viewport Info",
    description = "Display viewport dimensions and scroll position",
    action = "show_viewport",
    contexts = {"normal"},
    callback = function()
        local vp = editor.get_viewport()

        if vp then
            local msg = string.format("Viewport: %dx%d, top_byte=%d, left_col=%d",
                vp.width, vp.height, vp.top_byte, vp.left_column
            )
            editor.set_status(msg)
        else
            editor.set_status("No viewport info available")
        end
    end
})

editor.set_status("Buffer Query Demo plugin loaded! Try the 'Query Demo' commands.")

File: plugins/examples/buffer_query_demo.ts
/// <reference path="../../types/fresh.d.ts" />

/**
 * Buffer Query Demo Plugin
 * Demonstrates the buffer query APIs in Phase 2
 */

// Show buffer info
globalThis.show_buffer_info_demo = function(): void {
  const bufferId = editor.getActiveBufferId();
  const info = editor.getBufferInfo(bufferId);

  if (info) {
    const msg = `Buffer ${info.id}: ${info.path || "[No Name]"} (${
      info.modified ? "modified" : "saved"
    }, ${info.length} bytes)`;
    editor.setStatus(msg);
  } else {
    editor.setStatus("No buffer info available");
  }
};

editor.registerCommand(
  "Query Demo: Show Buffer Info",
  "Display information about the current buffer",
  "show_buffer_info_demo",
  "normal"
);

// Show cursor position with selection info
globalThis.show_cursor_info_demo = function(): void {
  const cursor = editor.getPrimaryCursor();

  if (cursor) {
    let msg: string;
    if (cursor.selection) {
      msg = `Cursor at ${cursor.position}, selection: ${cursor.selection.start}-${cursor.selection.end} (${
        cursor.selection.end - cursor.selection.start
      } chars)`;
    } else {
      msg = `Cursor at byte position ${cursor.position} (no selection)`;
    }
    editor.setStatus(msg);
  } else {
    editor.setStatus("No cursor info available");
  }
};

editor.registerCommand(
  "Query Demo: Show Cursor Position",
  "Display cursor position and selection info",
  "show_cursor_info_demo",
  "normal"
);

// Count all cursors (multi-cursor support)
globalThis.count_cursors_demo = function(): void {
  const cursors = editor.getAllCursors();
  editor.setStatus(`Active cursors: ${cursors.length}`);
};

editor.registerCommand(
  "Query Demo: Count All Cursors",
  "Display the number of active cursors",
  "count_cursors_demo",
  "normal"
);

// List all buffers
globalThis.list_all_buffers_demo = function(): void {
  const buffers = editor.listBuffers();
  let modifiedCount = 0;

  for (const buf of buffers) {
    if (buf.modified) {
      modifiedCount++;
    }
  }

  editor.setStatus(`Open buffers: ${buffers.length} (${modifiedCount} modified)`);
};

editor.registerCommand(
  "Query Demo: List All Buffers",
  "Show count of open buffers",
  "list_all_buffers_demo",
  "normal"
);

// Show viewport info
globalThis.show_viewport_demo = function(): void {
  const vp = editor.getViewport();

  if (vp) {
    const msg = `Viewport: ${vp.width}x${vp.height}, top_byte=${vp.top_byte}, left_col=${vp.left_column}`;
    editor.setStatus(msg);
  } else {
    editor.setStatus("No viewport info available");
  }
};

editor.registerCommand(
  "Query Demo: Show Viewport Info",
  "Display viewport dimensions and scroll position",
  "show_viewport_demo",
  "normal"
);

editor.setStatus("Buffer Query Demo plugin loaded! Try the 'Query Demo' commands.");
editor.debug("Buffer Query Demo plugin initialized (TypeScript version)");
